src/historywork/CheckSingleLedgerHeaderWork.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "historywork/CheckSingleLedgerHeaderWork.h"
+#include "history/FileTransferInfo.h"
+#include "history/HistoryArchive.h"
+#include "history/HistoryManager.h"
+#include "historywork/GetAndUnzipRemoteFileWork.h"
+#include "util/Logging.h"
+#include "util/TmpDir.h"
+#include "xdrpp/printer.h"
+
+namespace stellar
+{
+
+CheckSingleLedgerHeaderWork::CheckSingleLedgerHeaderWork(
+    Application& app, std::shared_ptr<HistoryArchive> archive,
+    LedgerHeaderHistoryEntry const& expected)
+    : Work(app,
+           fmt::format("check-single-ledger-header-{}",
+                       expected.header.ledgerSeq),
+           BasicWork::RETRY_NEVER)
+    , mArchive(archive)
+    , mExpected(expected)
+    , mCheckSuccess(
+          app.getMetrics().NewMeter({"history", "check", "success"}, "event"))
+    , mCheckFailed(
+          app.getMetrics().NewMeter({"history", "check", "failure"}, "event"))
+{
+}
+
+CheckSingleLedgerHeaderWork::~CheckSingleLedgerHeaderWork()
+{
+}
+
+void
+CheckSingleLedgerHeaderWork::onSuccess()
+{
+    mCheckSuccess.Mark();
+    Work::onSuccess();
+}
+
+void
+CheckSingleLedgerHeaderWork::onFailureRaise()
+{
+    mCheckFailed.Mark();
+    Work::onFailureRaise();
+}
+
+void
+CheckSingleLedgerHeaderWork::doReset()
+{
+    mGetLedgerFileWork.reset();
+    mDownloadDir =
+        std::make_unique<TmpDir>(mApp.getTmpDirManager().tmpDir(getName()));
+    uint32_t checkpoint = mApp.getHistoryManager().checkpointContainingLedger(
+        mExpected.header.ledgerSeq);
+    mFt = std::make_unique<FileTransferInfo>(
+        *mDownloadDir, HISTORY_FILE_TYPE_LEDGER, checkpoint);
+}
+
+BasicWork::State
+CheckSingleLedgerHeaderWork::doWork()
+{
+    if (!mGetLedgerFileWork)
+    {
+        CLOG_INFO(History, "Downloading ledger checkpoint {} from archive {}",
+                  mFt->baseName_gz(), mArchive->getName());
+        mGetLedgerFileWork = addWork<GetAndUnzipRemoteFileWork>(
+            *mFt, mArchive, BasicWork::RETRY_NEVER);
+        return State::WORK_RUNNING;
+    }
+    else if (!mGetLedgerFileWork->isDone())
+    {
+        return State::WORK_RUNNING;
```

Comment:
This doesn't look right, we need to return WAITING if mGetLedgerFileWork is waiting, otherwise this work will busy loop. So I think here you want to return `mGetLedgerFileWork->getState()`.

Reply:
Fixed.